libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/VoiceInstructionsFilePlayer.kt

```diff
@@ -0,0 +1,129 @@
+import java.io.IOException
+
+/**
+ * Online implementation of [VoiceInstructionsPlayer].
+ * Will retrieve synthesized speech mp3s from Mapbox's API Voice.
+ * @property context Context
+ * @property accessToken String
+ * @property language [Locale] language (ISO 639)
+ */
+internal class VoiceInstructionsFilePlayer(
+    private val context: Context,
+    private val accessToken: String,
+    private val language: String
+) : VoiceInstructionsPlayer {
+
+    private var mediaPlayer: MediaPlayer? = null
+    private var volumeLevel: Float = DEFAULT_VOLUME_LEVEL
+    private var clientCallback: VoiceInstructionsPlayerCallback? = null
+    private var currentPlay: SpeechState.ReadyToPlay? = null
+
+    /**
+     * Given [SpeechState.ReadyToPlay] [Announcement] the method will play the voice instruction.
+     * If a voice instruction is already playing or other announcement are already queued,
+     * the given voice instruction will be queued to play after.
+     * @param state SpeechState Play Announcement object including the announcement text
+     * and optionally a synthesized speech mp3.
+     * @param callback
+     */
+    override fun play(state: SpeechState.ReadyToPlay, callback: VoiceInstructionsPlayerCallback) {
+        clientCallback = callback
+        check(currentPlay == null) {
+            "Only one announcement can be played at a time."
+        }
+        currentPlay = state
+        val file = state.announcement.file
+        if (file != null && file.canRead()) {
+            play(file)
+        }
+    }
+
+    /**
+     * The method will set the volume to the specified level from [SpeechState.Volume].
+     * @param state SpeechState Volume level.
+     */
+    override fun volume(state: SpeechState.Volume) {
+        volumeLevel = state.level
+        setVolume(volumeLevel)
+    }
+
+    /**
+     * Clears any announcements queued.
+     */
+    override fun clear() {
+        resetMediaPlayer(mediaPlayer)
+    }
+
+    /**
+     * Releases the resources used by the speech player.
+     * If called while an announcement is currently playing,
+     * the announcement should end immediately and any announcements queued should be cleared.
+     */
+    override fun shutdown() {
+        clear()
+        volumeLevel = DEFAULT_VOLUME_LEVEL
+    }
+
+    private fun play(instruction: File) {
+        try {
+            FileInputStream(instruction).use { fis ->
+                mediaPlayer = MediaPlayer().apply {
+                    setDataSource(fis.fd)
+                    prepareAsync()
+                }
+                setVolume(volumeLevel)
+                addListeners()
+            }
+        } catch (ex: FileNotFoundException) {
+            donePlaying(mediaPlayer)
+        } catch (ex: IOException) {
+            donePlaying(mediaPlayer)
+        }
+    }
+
+    private fun addListeners() {
+        mediaPlayer?.run {
+            setOnErrorListener { _, _, _ ->
+                false
```

Comment:
```suggestion
                Log.e()
                onDonePlaying()
```

Reply:
`donePlaying` is not necessary, returning `false` will cause the `OnCompletionListener` to be called.